cirq-core/cirq/circuits/frozen_circuit.py

```diff
@@ -136,6 +136,11 @@ def are_all_measurements_terminal(self) -> bool:
 
     # End of memoized methods.
 
+    def _is_measurement_(self) -> bool:
+        if self._has_measurements is None:
+            return NotImplemented
```

Comment:
As noted above - if `_has_measurements` is not set, this should set it.

Reply:
Calculating `has_measurement` from scratch can be very costly. `_is_measurement_` cannot be expected to do the full circuit decomposition inside it, the protocol should handle it in the case where `_is_measurement_` does not have an efficient answer.

Reply:
This is why I'm concerned: for `fc = FrozenCircuit(...)` if someone invokes `cirq.is_measurement(fc)`, then later invokes `fc.has_measurement()`, the result will not be memoized for the `has_measurement` call.

Ideally, something like this would work:
```
def _is_measurement_(self) -> bool:
    if self._has_measurements is None:
        self._has_measurements = cirq.is_measurement(self.unfreeze())
    return self._has_measurements
```
Deferring `is_measurement` handling to the `Circuit` class prevents infinite recursion in the `_is_measurement_from_magic_method` step of the protocol.

Reply:
Updated to enable memoization. This comes with the downside that a `CircuitOperation`, once encountered, will now be immediately fully evaluated regardless of how much decomposing it needs. I was aesthetically hesitant to go this route since it deviates from a pure breadth-first search for a measurement. But memoizing the expensive `is_measurement` call is definitely a net win and worth it.

The `is_measurement` protocol might slow down a bit for circuits with many non-measurement terminal `CircuitOperations`, but these will probably be rare.

Reply:
Loss of BFS **is** unfortunate...I have a nagging feeling that there's a clean solution that allows BFS to coexist with proper memoization, but the existing tangle of methods and expected behavior make implementing that solution infeasible :/